### PR TITLE
poplar1: Partially implement the VDAF

### PR DESCRIPTION
**Pull request by cjpatton**
Partially addresses #141.

Begin implementing Poplar1. This commit includes the core logic fof the VDAF algorithms, but lacks the serialization code.

### CODE REVIEW

File: src/vdaf/poplar1.rs [hunk 1]
```diff
@@ -204,47 +345,528 @@ impl<P: Prg<L>, const L: usize> Aggregator<L> for Poplar1<P, L> {
...
+        level.to_vec(),
+        prefix_count.to_vec(),
+    ]
+    .into_iter()
+    .chain(agg_param.prefixes.iter().map(|_prefix| {
+        // XXX Encode each prefix as required by the spec.
+        //
+        // TODO(cjpatton) spec: Consider not encoding the prefixes here.
+        b"XXX".to_vec()
+    }));
```
Comment (cjpatton) [inline src/vdaf/poplar1.rs:663]: This is the last bit that's needed before this is ready to review.

**Review by tgeoghegan — COMMENTED** [review]
Looks sound, I just have a bunch of nits. Also one cross-cutting nits: I'm concerned about the number of TODO comments in here, because it's too easy to lose track of them and either leave them behind after they're addressed, or just never deal with them. I'm especially concerned for those items that require changes to the VDAF spec. Could those TODOs be promoted to issues on this repo or VDAF? Even just one issue with a checklist of items you want to address in subsequent PRs would be nice.

File: src/vdaf/poplar1.rs [hunk 2]
```diff
@@ -62,7 +70,22 @@ impl<P, const L: usize> ParameterizedDecode<Poplar1<P, L>> for Poplar1PublicShar
 /// Poplar1 input share. Comprised of an IDPF key share and the correlated randomness used to
 /// compute the sketch during preparation.
 #[derive(Debug, Clone)]
-pub struct Poplar1InputShare<const L: usize> {}
+pub struct Poplar1InputShare<const L: usize> {
+    /// IDPF key share.
+    idpf_key: Seed<L>,
+
+    /// Seed used to generate the Aggregator's share of the randomness used in the first part
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply src/vdaf/poplar1.rs:77]: If it's incorrect to describe it as the seed for "correlation randomness" in prose, then presumably the field name `corr_seed` is wrong.

File: src/vdaf/poplar1.rs [hunk 3]
```diff
@@ -204,47 +348,524 @@ impl<P: Prg<L>, const L: usize> Aggregator<L> for Poplar1<P, L> {
...
+    let mut verify_prng = init_prng::<_, _, _, _, L>(P::init(verify_key), info_suffix_parts);
+
+    let mut out_share = Vec::with_capacity(agg_param.prefixes.len());
+    let mut sketch_share = vec![
+        corr_prng.get(), // a_share
+        corr_prng.get(), // b_share
+        corr_prng.get(), // c_share
+    ];
+
+    let mut idpf_eval_cache = RingBufferCache::new(agg_param.prefixes.len());
```
[the remark this answers is not in the document]
Comment (cjpatton) [reply src/vdaf/poplar1.rs:675]: Yeah I noticed that `NoCache` actulaly seems to perform better. I've left a `TODO(divergentdave)` for you to take a crack at it later :)

A couple of things to keep in mind:
1. We might require in the spec that the candidate prefixes are stored in lexicographic order. (This makes it easier to check for repeats, which is something we know we have to do.) Might that be helpful here?
2. The next PR will have benchmarks to run against.

File: src/vdaf/poplar1.rs [hunk 4]
```diff
@@ -204,47 +356,506 @@ impl<P: Prg<L>, const L: usize> Aggregator<L> for Poplar1<P, L> {
...
+    // TODO(cjpatton) spec: Consider not encoding the prefixes here.
+    let mut verify_prng = init_prng::<_, _, _, _, L>(
+        P::init(verify_key),
+        [
+            DST_VERIFY_RANDOMNESS,
+            // TODO(cjpatton) spec: Drop length prefix if we decide to fix the nonce length.
+            &[nonce_len],
+            nonce,
+            &level,
+        ],
```
Comment (cjpatton) [inline src/vdaf/poplar1.rs:664]: Reviewer note: This value has changed to be more aligned with what I expect to land in VDAF-04. We decided yesterday that we don't need to implement VDAF-03 anyway.

File: src/vdaf/poplar1.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (divergentdave) [reply src/vdaf/poplar1.rs:675]: I was thinking that if you gave up on caching between different aggregation parameters, and were optimizing for very wide candidate prefix sets, lexicographic order means you could do pretty well with a fixed-size cache of one seed per level.

File: src/vdaf/poplar1.rs [hunk 5]
```diff
@@ -83,11 +106,36 @@ impl<'a, P, const L: usize> ParameterizedDecode<(&'a Poplar1<P, L>, usize)>
 
 /// Poplar1 preparation state.
 #[derive(Clone, Debug)]
-pub struct Poplar1PrepareState();
+pub struct Poplar1PrepareState(PrepareStateVar);
+
+#[derive(Clone, Debug)]
+enum PrepareStateVar {
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply src/vdaf/poplar1.rs:115]: That works for me.

File: src/vdaf/poplar1.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply src/vdaf/poplar1.rs:675]: cool, let's give it a try!

File: src/vdaf/poplar1.rs [hunk 6]
```diff
@@ -204,47 +348,524 @@ impl<P: Prg<L>, const L: usize> Aggregator<L> for Poplar1<P, L> {
...
+    auth: F,
+) -> ([F; 2], [F; 2]) {
+    let two = F::from(2);
+    let a = corr_prng_0.get() + corr_prng_1.get();
+    let b = corr_prng_0.get() + corr_prng_1.get();
+    let c = corr_prng_0.get() + corr_prng_1.get();
+    let x = -two * a + auth;
+    let y = a * a + b - a * auth + c;
+    let corr_0 = [prng.get(), prng.get()]; // A
+    let corr_1 = [x - corr_0[0], y - corr_0[1]]; // B
```
[the remark this answers is not in the document]
Comment (tgeoghegan) [reply src/vdaf/poplar1.rs]: I think it'd make it easier to map the implementation to the spec and the paper, so it's worth throwing an `#[allow(whatever)]` in.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
The change LGTM modulo one or two naming things, but I'd still like to see the TODOs reified into issues per https://github.com/divviup/libprio-rs/pull/434#pullrequestreview-1276192743

File: src/vdaf/poplar1.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply src/vdaf/poplar1.rs:675]: Moved to https://github.com/divviup/libprio-rs/issues/441.

File: src/vdaf/poplar1.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply src/vdaf/poplar1.rs]: Done, here and below.

**Comment by cjpatton** [comment]
> The change LGTM modulo one or two naming things, but I'd still like to see the TODOs reified into issues per [#434 (review)](https://github.com/divviup/libprio-rs/pull/434#pullrequestreview-1276192743)

Which naming things?

Filed:
https://github.com/divviup/libprio-rs/issues/441
https://github.com/divviup/libprio-rs/issues/442

**Comment by tgeoghegan** [comment]
> Which naming things?
Just `Var -> Variant`, which you already got 

> Filed: #441 #442
> 
> EDIT: I left the "spec TODOs" there because the issues would be filed against cfrg/draft-irtf-cfrg-vdaf.

Yeah, those are the ones that I'm most worried about because they're the least actionable from this project. Any reason not to open some issues on VDAF?

**Comment by cjpatton** [comment]
> Yeah, those are the ones that I'm most worried about because they're the least actionable from this project. Any reason not to open some issues on VDAF?

https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/134

**Review by tgeoghegan — APPROVED** [review]
Thanks for bearing with me. LGTM!

**Review by divergentdave — APPROVED** [review]

**Comment by cjpatton** [comment]
Squashed.

**Pull request merged**